packages/lab/src/slider/internal/utils.ts

```diff
@@ -1,158 +1,131 @@
   min: number,
   max: number,
+  step: number,
+  clientX: number,
 ) => {
-  const newValue = [...oldValue];
-  if (valueItem < oldValue[index]) {
-    const constraint = index === 0 ? min : newValue[index - 1];
-    newValue[index] = Math.max(constraint, valueItem);
-  } else {
-    const constraint =
-      index === newValue.length - 1 ? max : newValue[index + 1];
-    newValue[index] = Math.min(constraint, valueItem);
+  const { width, x } = trackRef.current!.getBoundingClientRect();
+  const localX = clientX - x;
+  const normaliseBetweenValues = (localX / width) * (max - min) + min;
+  const roundedToStep = roundToStep(normaliseBetweenValues, step);
+  const decimals = countDecimalPlaces(step);
+  const rounded = Number(roundedToStep.toFixed(decimals));
+  const value = clampValue(rounded, [min, max]);
+  return value;
+};
+
+export const setValue = (
+  value: SliderValue,
+  newValue: number,
+  index: ThumbIndex,
+  onChange: SliderChangeHandler,
+) => {
+  if (value.length === 2) {
+    const newValueArray = [...value];
+    newValueArray.splice(index, 1, newValue);
+    onChange(newValueArray as SliderValue);
+    return;
+  }
+  onChange([newValue]);
+};
+
+export const roundToStep = (value: number, step: number) =>
+  Math.round(value / step) * step;
+
+export const clampValue = (value: number, [min, max]: number[]) => {
+  if (value > max) {
+    return max;
+  }
+  if (value < min) {
+    return min;
   }
-  return newValue;
+  return value;
+};
+
+export const getPercentage = (min: number, max: number, value: number) => {
+  const percentage = ((value - min) / (max - min)) * 100;
+  return Math.min(Math.max(percentage, 0), 100);
 };
 
-const updateValueItemPushable = (
-  oldValue: number[],
-  index: number,
-  valueItem: number,
+export const getPercentageDifference = (
   min: number,
   max: number,
-  pushDistance: number,
+  value: number[],
 ) => {
-  const newValue = [...oldValue];
-  newValue[index] = valueItem;
-  if (valueItem < oldValue[index]) {
-    for (let i = index - 1; i >= 0; --i) {
-      if (newValue[i + 1] - newValue[i] < pushDistance) {
-        newValue[i] = newValue[i + 1] - pushDistance;
-      } else {
-        break;
-      }
-    }
-    const distToMin = newValue[0] - min;
-    if (distToMin < 0) {
-      for (let i = index; i >= 0; --i) {
-        newValue[i] -= distToMin;
-      }
-    }
-  } else {
-    for (let i = index + 1; i < newValue.length; ++i) {
-      if (newValue[i] - newValue[i - 1] < pushDistance) {
-        newValue[i] = newValue[i - 1] + pushDistance;
-      } else {
-        break;
-      }
-    }
-    const distToMax = max - newValue[newValue.length - 1];
-    if (distToMax < 0) {
-      for (let i = index; i < newValue.length; ++i) {
-        newValue[i] += distToMax;
-      }
-    }
-  }
-  return newValue;
+  const valueDiff = value[1] - value[0];
+  const percentage = ((valueDiff - min) / (max - min)) * 100;
+  return `${Math.min(Math.max(percentage, 0), 100)}%`;
```

Comment:
I feel like the percentage returned by the default, diff and offset functions should be consistent